src/utils.ts

```diff
@@ -157,8 +157,8 @@ export function parseIndexOptions(indexSpec: IndexSpecification): IndexOptions {
     // {location:'2d', type:1}
     keys = Object.keys(indexSpec);
     keys.forEach(key => {
-      indexes.push(key + '_' + indexSpec[key]);
-      fieldHash[key] = indexSpec[key];
+      indexes.push(key + '_' + (indexSpec as any)[key]);
```

Comment:
The type narrowing in this function isn't good, adding readonly arrays as a possible input made this section of code think indexSpec is the entire union again instead of what it was previously narrowed too. 

I could refactor but how about we do that in a follow up? (Thanks for calling it out I planned for the ticket option if that's okay with you)

Reply:
I'm fine with a follow up ticket

Reply:
https://jira.mongodb.org/browse/NODE-3517